### ISSUE TITLE
Output PSG/FM-PAC/SCC as MIDI file

### ISSUE THREAD

**Issue by taupter** [openMSX/openMSX#1039]
openMSX has a fantastic time-perfect way to export individual audio channels to WAV files that is very handy when replaying MSX's beloved tunes. The only thing that's missing is the fact that although WAV files are useful, they aren't adequate to editing notes or changing instruments, or studying how the actual note sequences are.
MIDI is a battle-tested standard that is considered an universal way to store and represent actual musical notes and note-changing messages. Being able to output music data as MIDI messages to a file would be a very nice feature to have to a musician. openMSX already does so to MSX's MIDI applications, but for non-MIDI music data, as for PSG, FM-PAC and SCC, openMSX currently doesn't have a meaningful way to export those melodies in a musician-friendly way.
There are at least another MSX emulator, fMSX, that is able to, at least partially, do what this issue is all about (parameter -logsnd), but their implementation is incomplete, as no time information is stored, only the raw notes in sequence, making it more useful as a debug output or something that can be piped do a MIDI recording program as a virtual device. Information about some nuances of the music aren't sent too (portamento, patch change and others).
It's a good opportunity for openMSX to have such an useful feature, and even implement time-accurate recording of it (using 960 ticks per beat and 240 bpm can give a 0.260416ms resolution), to be rendered independent of the actual speed of the emulation.
To a musician it's much easier to quantize and time-stretch MIDI data to extract in-time music than to convert audio files to note data, or type note data based on what he's listening. A MIDI file with notes will be extremely handy.
Please consider implementing this feature.

**Comment by MBilderbeek** [comment]
Good idea! But I think also quite some work. Help would be greatly appreciated to implement this...!

**Comment by taupter** [comment]
Well, I can code, but I'd use some guidance if provided. I programmed Assembly on the real hardware and have experience with embedded platforms, but not with emulation. Doing a very wild overview of the problem I'd look at the code that interfaces with the AY8910 at first (simpler, three tone channels and a noise one) and watch so changes to the 12-bit register according to page 47 of the AY-8910 programming manual (http://dev-docs.atariforge.org/files/GI_AY-3-8910_Feb-1979.pdf) and generate some debug output to console or to a file. After that starting implementing a real SMF file writer (independent module) and do a translation middleware to convert changes to registers to actual MIDI notes, volume changes and whatnot. Having it done and running smooth it would be time to implement SCC, SCC+ and then YM2413. But I'll need to stick around in Freenode and bug you once in a while for directions.

**Comment by mthuurne** [comment]
It might make sense to implement the SMF writer first, as a MIDI pluggable. You could then test it by connecting it to an MSX MIDI interface in openMSX.

I guess you could have a generic freq + amplitude to MIDI event conversion routine that is shared between all sound chips.

The code to extract freq + amplitude for each chip could either be done in C++ or in Tcl. Look at the existing "sccviewer" and "psg2scc" toys scripts for examples of how to read sound chips in Tcl using the debug interfaces.

It would be useful to take a look at [MIDI-PAC](http://worp3.com/MainPage.php?id=Projects&subid=midipac), a cartridge for MSX that translates MSX-MUSIC events to MIDI. Maybe emulate the MIDI-PAC itself or have an option to use its instrument configurations.
